### PR TITLE
 Annotated<T, ...> interface to enhance in-code block and parameter field documentation

### PR DESCRIPTION
**Pull request by RalphSteinhagen**
This commit introduces a robust, transparent Annotated<T, ...> type wrapper interface, paving the way for user-friendly, embedded, in-code documentation. Here's how it works and how it benefits our development process:

**Transparent Type Wrapper:**
`Annotated<T, ...>` wraps around any type, preserving its behaviour while augmenting it with added meta information. This information can range from documentation and unit info to UI behaviour controls and beyond. The variable type argument accepts any meta-information type, with built-in support for optional `Doc<"...">` and `Unit<"...">` that allow for more detailed (e.g. using mark-down) documentation of blocks, its parameters, and their physical (e.g. SI) units. This flexibility enables user extendability, creating a flexible interface that adapts to diverse needs.

**Streamlined Documentation:**
Embedding documentation directly in the code removes the need for separate documentation files or automated generation tools. This approach fosters a single source of truth for block definitions, aligning documentation with the actual source code and enhancing the maintainability of the project and OOT modules.

**Enabling UI Behaviour:**
The annotations can drive specific UI behaviours, such as the in-UI detailed block or parameter documentation, highlighting or suppressing specific parameters, and more.

**Example Usage:**
The `TestBlock` class showcases the usage of this interface. It uses `Doc` and `Unit` to annotate the class and the `scaling_factor` field, and `Visible` which may be used to control UI visibility.

```cpp
using TestBlockDoc = Doc<R""(
#TestBlock

some test doc documentation -- may use mark down, references etc. -- and can be read-out programmatically 

// optional future extension: 
// use existing input/output port information and constraints for additional documentation
)"">;

template<typename T>
struct TestBlock : public node<TestBlock<T>, TestBlockDoc> {
    IN<T>  in;
    OUT<T> out;
    A<T, "scaling factor", Visible, Doc<"y = a * x">, Unit<"As">> scaling_factor = static_cast<T>(1); 
    A<std::string, "context information", Visible>                context;
    ...
};
```

and some snippets from the corresponding [unit-test](https://github.com/fair-acc/graph-prototype/blob/92af03690a96c8e228276aef30f63527fde2a28a/test/qa_settings.cpp#L310-L330):
```cpp
graph flow_graph;
TestBlock<float> &block = flow_graph.make_node<TestBlock<float>>();
expect(eq(block.description(), std::string_view(TestBlockDoc::value)));
expect(block.scaling_factor.visible());
expect(eq(block.scaling_factor.description(), std::string_view{ "scaling factor" }));
expect(eq(block.scaling_factor.unit(), std::string_view{ "As" }));
expect(eq(block.context.unit(), std::string_view{ "" }));
expect(block.context.visible());

block.scaling_factor = 42.f; // test wrapper assignment operator
expect(block.scaling_factor == 42) << "the answer to everything failed -- equal operator";
expect(eq(block.scaling_factor.value, 42)) << "the answer to everything failed -- by value";
expect(eq(block.scaling_factor, 42)) << "the answer to everything failed -- direct";
```

In summary, this commit takes a significant leap towards improving the user-friendliness and maintainability of our project by embedding documentation and other metadata directly into the code.

### **Important:** 
This pre-version is open for feedback, comments, and suggestions. 

The hope is to iterate over, for example, naming, additional features, required meta-info types, etc. to improve the GNU Radio 4.0 development user experience.

### CODE REVIEW

**Review by ivan-cukic — COMMENTED** [review]
Ok, with a note from the chat that using tuple is an overkill for find_type.

p.s. Can you add `#include <tuple>` to utils.hpp - we don't include it, and it is used in quite a few places in the file

**Comment by RalphSteinhagen** [comment]
> Ok, with a note from the chat that using tuple is an overkill for find_type.
> 
> p.s. Can you add `#include <tuple>` to utils.hpp - we don't include it, and it is used in quite a few places in the file

Thanks and I agree. Using `std::tuple<>` was by accident. :grimacing: 

Will refactor and add also the include.

**Comment by RalphSteinhagen** [comment]
> But the only available alternative I see is to provide descriptions and everything inline in annotated's template parameters which is still possible and also not so nice from a readability perspective for longer doc-strings.

Just for giving a visual example of  how the inlining would like:

```cpp
template<typename T>
struct TestBlock : public node<TestBlock<T>, 
Doc<R""(
some test doc documentation -- may use mark down, references etc. -- and can be read-out programmatically 

// optional future extension: 
// use existing input/output port information and constraints for additional documentation
)"">> {
    IN<T>  in;
    OUT<T> out;
    A<T, "scaling factor", Visible, Doc<"y = a * x">, Unit<"As">> scaling_factor = static_cast<T>(1); 
    A<std::string, "context information", Visible>                context;
    ...
};
```
(N.B. I omitted the block name because this is trivially available through compile-time reflection)

@wirew0rm as you mentioned this should be discussed with a wider audience ...

**Comment by RalphSteinhagen** [comment]
Added `BlockingIO` and `SupportedTypes<..>` meta-info types and adjusted the node documentation generation accordingly to illustrate their use cases better.

```cpp
template<typename T>
struct TestBlock : public node<TestBlock<T>, BlockingIO, SupportedTypes<float, double>,
Doc<R""(
some test doc documentation -- may use mark down, references etc. -- and can be read-out programmatically 

// optional future extension: 
// use existing input/output port information and constraints for additional documentation
)"">> {
    IN<T>  in;
    OUT<T> out;
    A<T, "scaling factor", Visible, Doc<"y = a * x">, Unit<"As">> scaling_factor = static_cast<T>(1); 
    A<std::string, "context information", Visible>                context;
    ...
};
```

generates the following markdown description via `node_description<TestBlock<float>>()` (of course: documentation format is subject to improvement and changes, feel free to make proposals):



# fair::graph::setting_test::TestBlock<float>

some test doc documentation -- may use mark down, references etc. -- and can be read-out programmatically 

// optional future extension: 
// use existing input/output port information and constraints for additional documentation

**BlockingIO**
_i.e. potentially non-deterministic/non-real-time behaviour_

**supported data types:** 0:float 1:double 
**Parameters:**
float      scaling_factor       - annotated info: scaling factor unit: [As] documentation: y = a * x
std::basic_string<Elem, Traits, Alloc> context              - annotated info: context information unit: [] documentation: 
signed int n_samples_max       
float      sample_rate         

~~Ports:~~
tbd.

**Review by mattkretz — CHANGES_REQUESTED** [review]
Looks good in general. But see inline comments.

File: include/annotated.hpp [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+ * @brief Annotates node, indicating to calling schedulers that it may block due IO.
+ */
+struct BlockingIO {};
+
+/**
+ * @brief Annotates templated node, indicating which port data types are supported.
+ */
+template<typename... Ts>
+struct SupportedTypes {};
+
```
Comment (ivan-cukic) [inline include/annotated.hpp:70]: The approach of adding supported type to node's parameter list has a serious limitation that I realized just yesterday working on the plugin system - it is not accessible until a node is instantiated. But we can't create a node as we don't know the `T` for which we can create it.

File: include/annotated.hpp [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        value = std::move(value_);
+        return *this;
+    }
+
+    inline constexpr
+    operator T &() noexcept {
+        return value;
+    }
+
+    inline constexpr operator const T &() const noexcept { return value; }
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply include/annotated.hpp:114]: this is according to the existing clang-format definitions ... close for now since this is not functional.

File: include/annotated.hpp [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+/**
+ * @brief Annotated is a template class that acts as a transparent wrapper around another type.
+ * It allows adding additional meta-information to a type, such as documentation, unit, and visibility.
+ * The meta-information is supplied as template parameters.
+ */
+template<typename T, fair::meta::fixed_string description_ = "", typename... Arguments>
+struct Annotated {
+    using value_type = T;
+    T value{};
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply include/annotated.hpp:87]: Was intentional to not have uninitialised variables ... but I see your point will remove.

File: include/annotated.hpp [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+using EmptyUnit = Unit<"">; // nomen-est-omen
+
+template<typename T>
+struct is_unit : std::false_type {};
+
+template<fair::meta::fixed_string N>
+struct is_unit<Unit<N>> : std::true_type {};
+
+template<typename T>
+concept UnitType = is_unit<T>::value;
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply include/annotated.hpp:48]: Thanks for pointing out the naming issue.  We already have `static_assert(fair::meta::is_instantiation_of<SupportedTypes<float, double>, SupportedTypes>);`. 

Will change and base the check on that... it's a cleaner solution and foregoes the is_... interfaces.

The issue with lengthening `Doc<T>` to `Documentation<T>` is that we want to use this also for fields which would blow-up the line-length a bit ... I would leave it for now since this can be exchanged at a later stage but will put this up for discussion also for the other GR developers.

File: include/annotated.hpp [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+template<typename T>
+struct is_supported_type_type : std::false_type {};
+
+template<typename... Ts>
+struct is_supported_type_type<SupportedTypes<Ts...>> : std::true_type {};
+
+using DefaultSupportedTypes = SupportedTypes<>;
+
+/**
+ * @brief Annotated is a template class that acts as a transparent wrapper around another type.
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply include/annotated.hpp:80]: Being 'transparent' is referring to the implicit conversion operator that returns a reference to `.value`. 
That would make it transparent, wouldn't it?

File: include/annotated.hpp (hunk 1 above)
Comment (RalphSteinhagen) [reply include/annotated.hpp:70]: need to think a bit more about this ... :thinking:

File: include/annotated.hpp [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+ * @brief Annotates templated node, indicating which port data types are supported.
+ */
+template<typename... Ts>
+struct SupportedTypes {};
+
+template<typename T>
+struct is_supported_type_type : std::false_type {};
+
+template<typename... Ts>
+struct is_supported_type_type<SupportedTypes<Ts...>> : std::true_type {};
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply include/annotated.hpp:75]: fixed to is_supported_types

File: include/annotated.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (mattkretz) [reply include/annotated.hpp:48]: What blows up the line-length is not pressing the return key :stuck_out_tongue_winking_eye:. This is fine:
```c++
struct TestBlock : public node<TestBlock<T>, BlockingIO,
                               Documentation<"some test doc documentation">,
                               SupportedTypes<float, double>> 
```
Once you start using `R""(...)""` you want an alias anyway, no?

It's simpler if users don't have to remember which names to abbreviate and which not. Just type (and auto-complete) full names.

File: test/qa_settings.cpp [hunk 7]
```diff
@@ -102,17 +102,22 @@ struct Source : public node<Source<T>> {
...
+    IN<T>  in;
+    OUT<T> out;
+    // parameters
+    A<T, "scaling factor", Visible, Doc<"y = a * x">, Unit<"As">> scaling_factor = static_cast<T>(1); // N.B. unit 'As' = 'Coulomb'
+    A<std::string, "context information", Visible>                context;
+    std::int32_t                                                  n_samples_max = -1;
+    float                                                         sample_rate   = 1000.0f;
+    std::vector<T>                                                vector_setting{ T(3), T(2), T(1) };
+    int                                                           update_count = 0;
+    bool                                                          debug        = false;
```
[the remark this answers is not in the document]
Comment (mattkretz) [reply test/qa_settings.cpp:124]: I commented on this because these lines show up in the diff even though only whitespace was changed. This is a huge downside of alignment, which is why I would turn those options off for clang-format. (Whether these lines were mine doesn't matter much. I followed the clang-format rules as defined in the project root. And now those rules start to bite.) For the sake of `git blame` and to reduce clutter in diffs I strongly prefer patches to not change formatting of unrelated lines. If re-formatting is necessary do it in a formatting-only commit.
